Batching with array

RCasatta opened the pull request:
Currently electrs support batching by separating requests via new lines, however, other impls support it also via json array  as explained in the cited issue. This add support for batching requests via json array

close https://github.com/Blockstream/esplora/issues/516

RCasatta committed:
add electrum raw test

RCasatta committed:
refactor out handling a single Value

RCasatta committed:
handle_value return result instead of sending

RCasatta committed:
handle batch requests as array

shesek commented on src/electrum/server.rs:
```diff
@@ -601,6 +569,55 @@ impl Connection {
+        &mut self,
+        cmd: Value,
+        empty_params: &Value,
+        start_time: Instant,
```
Shouldn't the start_time start ticking separately for each `handle_value` call?

RCasatta replied:
yes, it makes sense. Done in b8f4524cdbe8f280da9fd87d2a4ac136f630a30f

RCasatta committed:
timer tick separately for each rpc req

shesek commented:
Should we limit the number of requests per batch?

shesek commented on tests/electrum.rs:
```diff
@@ -137,3 +140,44 @@ fn test_electrum() -> Result<()> {
     Ok(())
 }
+
+/// Test the Electrum RPC server using an headless Electrum wallet
```
The comment here should be updated, "using a raw TCP socket" or something similar

RCasatta replied:
Done in 250fb9b33d990822f06c6b1ef552c0ebd137e93a

RCasatta commented:
> Should we limit the number of requests per batch?

Do we have limits for "new-line separated" batch requests? Is it different here? How much do you propose as the limit here?

shesek commented:
> Do we have limits for "new-line separated" batch requests? Is it different here?

We don't, but it is different here. New-lines separated requests are streamed, and individual responses are sent as they become available. We only hold up to 10 pending requests in the queue, then block and stop reading from the TCP socket until a pending request gets processed to free room (a bounded `mpsc::sync_channel`).

With array batching all the requests and all the response has to be buffered in-memory at once, which could use significant resources and be a DoS vector if there are no limits in place.

RCasatta commented:
> How much do you propose as the limit here

Do you think we can afford 20 as the limit for batch requests? I think it's common number because is used as gap limit...

RCasatta committed:
limit max limit of batch request via json array

shesek commented:
20 seems fine, yes :+1:

However it seems the tests are failing. The failure isn't directly related to the batching but to the TestRunner used to test it.

RCasatta committed:
ci: run on PR or push on new-index

RCasatta committed:
ignore test and execute singularly

At the moment we cannot concurrently tests because the server will
conflict with each other, this make the test_electrum_raw test ignored
by default, and add a test in CI specifically for this test.

RCasatta commented:
I didn't realize locally because i launched the test singularly, I think we don't support having multiple concurrent tests.
My solution is to make the test ignored by default and test it specifically.

RCasatta committed:
fix test comment

RCasatta commented:
Rebased on `new-index` and fixed forgotten comment

shesek commented:
> I think we don't support having multiple concurrent tests.

I tracked down the cause. It's not specific to the tests, it fails whenever `electrs` is run multiple times in the same process, even if not concurrently.

This appears to happen because the threads sending parallel RPC requests to bitcoind are being reused across different electrs instances. Each RPC thread has a `thread_local` var with a cached `Daemon` instance (and TCP socket). The thread initially initializes a `Daemon` connection to Instance A's bitcoind, then later when the same thread is reused for instance B but the cached connection to A is returned instead - causing the mismatch errors.

If you apply this patch:
```patch
diff --git a/src/daemon.rs b/src/daemon.rs
index 6808907..48498c7 100644
--- a/src/daemon.rs
+++ b/src/daemon.rs
@@ -493,9 +493,9 @@ impl Daemon {
                 // Store a local per-thread Daemon, each with its own TCP connection. These will
                 // get initialized as necessary for the `rpc_threads` pool thread managed by rayon.
                 thread_local!(static DAEMON_INSTANCE: OnceCell<Daemon> = OnceCell::new());
 
                 DAEMON_INSTANCE.with(|daemon| {
-                    daemon
-                        .get_or_init(|| self.retry_reconnect())
-                        .retry_request(&method, &params)
+                    let d = daemon.get_or_init(|| self.retry_reconnect());
+                    debug!("rpc thread {:?} of {} got {}", std::thread::current().id(), self.daemon_dir.display(), d.daemon_dir.display());
+                    d.retry_request(&method, &params)
                 })
             })
         })
```

And run:
```
RUST_LOG=debug cargo test electrum -- --nocapture --test-threads 1 2>&1 | grep 'rpc thread'
```

You'll see:

```
DEBUG rpc thread ThreadId(17) of /tmp/.tmp8eAz3c/regtest got /tmp/.tmp8eAz3c/regtest
...
DEBUG rpc thread ThreadId(17) of /tmp/.tmptlPtvs/regtest got /tmp/.tmp8eAz3c/regtest
```

Note how the second time its trying to get a `Daemon` connected to `tmptlPtvs`, but gets the previous `tmp8eAz3c` instead because it ran on the same thread.

I was expecting the threads not to be reused across electrs instances, because each is using [its own `rayon::ThreadPool`](https://github.com/Blockstream/electrs/blob/38d252537bcc01a4c6ade54961df191bc65aec85/src/daemon.rs#L328-L329). Can't quite figure out what's causing this. Any ideas?

RCasatta commented:
I am not sure how to handle this...

1) avoid the thread local in testing? 
2) Doing the raw test at the end of `test_electrum` so that not 2 RPC are started
3) Merge this (test ignored, launched singularly) and open an issue

shesek reviewed (dismissed):
tACK 250fb9b

Let's merge as-is for now with the new test launched singularly, I'll continue investigating the cause for the thread reuse after that.

RCasatta commented:
I had to rebase after #128

MASIHmirsali65 commented:
I am the original adoptive father of Bitcoin, please, without any reason or reason, open all the scams and tokens that are related to our IP and block the obstacles. Please get control as soon as possible.
